File: oceanmgr/src/db_oc.py
from dto_oc import BlockEarning
from ocean_web import EarningSnapshot

import sqlite3
from datetime import datetime

def db_setup_1(conn: sqlite3.Connection):
    cursor = conn.cursor()

    # Create table
    cursor.execute("""
        CREATE TABLE OC_BLOCK_EARN
        (Time INTEGER, BlockHash VARCHAR(100), Earning INTEGER, PoolFee INTEGER, TimeAddedFirst INTEGER, TimeUpdated INTEGER)
    """)
    cursor.execute("CREATE INDEX OcBlockEarnTime ON OC_BLOCK_EARN (Time)")

    cursor.execute("CREATE TABLE OC_EARN (Time INTEGER, Estimated INTEGER, AcctdUnpaid INTEGER, AcctdPaid INTEGER)")
    cursor.execute("CREATE INDEX OcEarnTime ON OC_EARN (Time)")

    # Commit changes and close connection
    conn.commit()


def insert_or_update_block_earning(cursor, earning: BlockEarning, now: datetime):
    cursor.execute("SELECT COUNT(*) FROM OC_BLOCK_EARN WHERE BlockHash = ?", (earning.block_hash,))
    rows = cursor.fetchall()
    if len(rows) >= 1:
        if len(rows[0]) >= 1:
            if rows[0][0] >= 1:
                # already present, update
                cursor.execute(
                    "UPDATE OC_BLOCK_EARN SET Time = ?, Earning = ?, PoolFee = ?, TimeUpdated = ? WHERE BlockHash = ?",
                    (earning.time, earning.earned_sats, earning.pool_fee, now, earning.block_hash,)
                )
                # print("updated")
                return
    # not present, insert
    cursor.execute("""
        INSERT INTO OC_BLOCK_EARN
        (Time, BlockHash, Earning, PoolFee, TimeAddedFirst, TimeUpdated)
        VALUES (?, ?, ?, ?, ?, ?)
    """, (earning.time, earning.block_hash, earning.earned_sats, earning.pool_fee, now, now,))
    # print("inserted")

def block_earnings_count(cursor) -> int:
    cursor.execute("SELECT COUNT(*) FROM OC_BLOCK_EARN")
    rows = cursor.fetchall()
    if len(rows) >= 1:
        if len(rows[0]) >= 1:
            return rows[0][0]
    return 0

def block_earnings_sum(cursor) -> int:
    cursor.execute("SELECT SUM(Earning) FROM OC_BLOCK_EARN")
    rows = cursor.fetchall()
    if len(rows) >= 1:
        if len(rows[0]) >= 1:
            return rows[0][0]
    return 0

def get_last_block(cursor: sqlite3.Cursor) -> BlockEarning:
    cursor.execute("""
        SELECT
        Time, BlockHash, Earning, PoolFee, TimeAddedFirst, TimeUpdated
        FROM OC_BLOCK_EARN
        ORDER BY Time DESC
        LIMIT 1
    """)
    rows = cursor.fetchall()
    if len(rows) >= 1:
        row = rows[0]
        if len(row) >= 4:
            return BlockEarning(row[0], row[1], row[2], row[3])
    return None

# Get the blocks after a certain time, oldest first.
# Old time is typically the time of the already processed last block.
def get_new_blocks(cursor: sqlite3.Cursor, old_time: int) -> list[BlockEarning]:
    cursor.execute("""
        SELECT
        Time, BlockHash, Earning, PoolFee, TimeAddedFirst, TimeUpdated
        FROM OC_BLOCK_EARN
        WHERE Time > ?
        ORDER BY Time ASC
    """, (old_time,))
    rows = cursor.fetchall()
    res = []
    for row in rows:
        if len(row) >= 4:
            res.append(BlockEarning(row[0], row[1], row[2], row[3]))
    return res


# Count the blocks after a certain time, oldest first.
# Old time is typically the time of the already processed last block.
def count_new_blocks(cursor: sqlite3.Cursor, old_time: int) -> int:
    cursor.execute("""
        SELECT COUNT(*)
        FROM OC_BLOCK_EARN
        WHERE Time > ?
    """, (old_time,))
    rows = cursor.fetchall()
    if len(rows) >= 1:
        if len(rows[0]) >= 1:
            return rows[0][0]
    return 0


def insert_earn_snapshot(conn, earns: EarningSnapshot):
    cursor = conn.cursor()

    cursor.execute(
        """
            INSERT INTO OC_EARN 
                (Time, Estimated, AcctdUnpaid, AcctdPaid)
            VALUES
                (?, ?, ?, ?)
        """,
        (earns.time, earns.estimated, earns.accounted_unpaid, earns.accounted_paid,)
    )

    cursor.close()
    conn.commit()

def get_last_snapshot(cursor) -> EarningSnapshot:
    cursor.execute("SELECT Time, Estimated, AcctdUnpaid, AcctdPaid FROM OC_EARN ORDER BY Time DESC")
    rows = cursor.fetchall()
    if len(rows) < 1:
        return None
    if len(rows[0]) < 4:
        return None
    return EarningSnapshot(int(rows[0][0]), int(rows[0][3]), int(rows[0][2]), int(rows[0][1]))


def get_last_snapshot_before(cursor, before_time: int) -> EarningSnapshot:
    cursor.execute("SELECT Time, Estimated, AcctdUnpaid, AcctdPaid FROM OC_EARN WHERE Time < ? ORDER BY Time DESC", (before_time,))
    rows = cursor.fetchall()
    if len(rows) < 1:
        return None
    if len(rows[0]) < 4:
        return None
    return EarningSnapshot(int(rows[0][0]), int(rows[0][3]), int(rows[0][2]), int(rows[0][1]))


def get_all_snapshots(cursor) -> dict[int, int]:
    cursor.execute("SELECT Time, Estimated, AcctdUnpaid, AcctdPaid FROM OC_EARN ORDER BY Time DESC")
    rows = cursor.fetchall()
    res = {}
    for row in rows:
        if len(row) >= 4:
            time = int(row[0])
            paid = int(row[3])
            total = paid + int(row[2]) + int(row[1])
            # print(time, total)
            res[time] = [total, paid]
    return res

File: paycalc-rust/src/db_oc.rs
use crate::dto_oc::BlockEarning;

use rusqlite::{Connection, Row};
use std::vec::Vec;
use std::error::Error;

fn blockearning_from_row(row: &Row) -> Result<BlockEarning, rusqlite::Error> {
    // println!("blockearning_from_row {0:?}", row);
    let w = BlockEarning::new(
        row.get(0)?,
        row.get::<_, String>(1)?,
        row.get(2)?,
        row.get(3)?,
    );
    // println!("blockearning_from_row {0}", w.block_hash);
    Ok(w)
}

pub fn get_new_blocks(conn: &Connection, old_time: u32) -> Result<Vec<BlockEarning>, Box<dyn Error>> {
    let query_str =
        "SELECT Time, BlockHash, Earning, PoolFee, TimeAddedFirst, TimeUpdated \
        FROM OC_BLOCK_EARN \
        WHERE Time > ?1 \
        ORDER BY Time ASC ";

    let mut stmt = conn.prepare(query_str)?;
    let vector = stmt.query_map([old_time], |row| blockearning_from_row(row))?
        .filter(|ber| ber.is_ok())
        .map(|ber| ber.unwrap())
        .collect::<Vec<BlockEarning>>();
    Ok(vector)
}

pub fn count_new_blocks(conn: &Connection, old_time: u32) -> Result<u32, Box<dyn Error>> {
    let mut stmt = conn.prepare(
        "SELECT COUNT(*) FROM OC_BLOCK_EARN WHERE Time > ?"
    )?;

    let res = stmt.query_one([old_time], |row| {
        Ok(row.get::<_, u32>(0))
    })??;
    Ok(res)
}


#[cfg(test)]
mod tests {
    use super::*;
    use rusqlite::Connection;

    fn db_setup_1(conn: &Connection) -> Result<(), Box<dyn Error>> {
        // Create table OC_BLOCK_EARN
        let _ = conn.execute(
            "CREATE TABLE OC_BLOCK_EARN \
            (Time INTEGER, BlockHash VARCHAR(100), Earning INTEGER, PoolFee INTEGER, TimeAddedFirst INTEGER, TimeUpdated INTEGER);",
            [],
        )?;
        Ok(())
    }

    fn create_test_db(conn: &Connection) -> Result<(), Box<dyn Error>> {
        // Create a test database with WORK table
        db_setup_1(&conn)?;
        conn.execute("INSERT INTO OC_BLOCK_EARN (Time, BlockHash, Earning, PoolFee, TimeAddedFirst, TimeUpdated) VALUES (1001, 'block_01', 11, 1, 1001, 1001);", [])?;
        conn.execute("INSERT INTO OC_BLOCK_EARN (Time, BlockHash, Earning, PoolFee, TimeAddedFirst, TimeUpdated) VALUES (1101, 'block_02', 22, 2, 1101, 1101);", [])?;
        Ok(())
    }

    #[test]
    fn test_count_new_blocks() -> Result<(), Box<dyn Error>> {
        let connection = Connection::open_in_memory()?;
        create_test_db(&connection)?;

        // Test our function
        let count = count_new_blocks(&connection, 900)?;
        assert_eq!(count, 2);

        Ok(())
    }

    #[test]
    fn test_get_new_blocks() -> Result<(), Box<dyn Error>> {
        let connection = Connection::open_in_memory()?;
        create_test_db(&connection)?;

        { // all
            let count = get_new_blocks(&connection, 900)?;
            assert_eq!(count.iter().len(), 2);
        }
        { // later ID
            let count = get_new_blocks(&connection, 1050)?;
            assert_eq!(count.iter().len(), 1);
        }
        { // ID just below a block
            let count = get_new_blocks(&connection, 1000)?;
            assert_eq!(count.iter().len(), 2);
        }
        { // ID just at a block
            let count = get_new_blocks(&connection, 1001)?;
            assert_eq!(count.iter().len(), 1);
        }
        { // ID just after a block
            let count = get_new_blocks(&connection, 1002)?;
            assert_eq!(count.iter().len(), 1);
        }

        Ok(())
    }
}
